source/core_mqtt_serializer.c

```diff
@@ -1706,9 +1706,9 @@ MQTTStatus_t MQTT_GetConnectPacketSize( const MQTTConnectInfo_t * pConnectInfo,
                     ( void * ) pPacketSize ) );
         status = MQTTBadParameter;
     }
-    else if( ( pConnectInfo->clientIdentifierLength == 0U ) || ( pConnectInfo->pClientIdentifier == NULL ) )
+    else if( ( ( pConnectInfo->clientIdentifierLength == 0U ) || ( pConnectInfo->pClientIdentifier == NULL ) ) && ( pConnectInfo->cleanSession == false ) )
     {
-        LogError( ( "Mqtt_GetConnectPacketSize() client identifier must be set." ) );
+        LogError( ( "Zero-length client identifier requires cleanSession=true per MQTT 3.1.1." ) );
         status = MQTTBadParameter;
     }
```

Comment:
We should also add this below check since on line `connectPacketSize += pConnectInfo->clientIdentifierLength + sizeof( uint16_t );` we are updating the length based only on the length provided by the user without checking whether the value itself is NULL. That will cause an issue.

@Ryan-CW-Code what do you think? I am open to combining the checks too. But I am currently out of office and can help once I am back.

```suggestion
        status = MQTTBadParameter;
    }
    else if( ( pConnectInfo->clientIdentifierLength == 0U ) ^ ( pConnectInfo->pClientIdentifier == NULL ) )
    {
        LogError( ( "Client ID length and value mismatch." ) );
        status = MQTTBadParameter;
    }
```